elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryCancelThread.java

```diff
@@ -0,0 +1,112 @@
+    public void run() {
+        cancelAsyncQuery();
+    }
+
+    /**
+     * This method cancels queries based on threshold.
+     */
+    protected void cancelAsyncQuery() {
+
+        try {
+
+            TransactionRegistry transactionRegistry = elide.getTransactionRegistry();
+            Map<UUID, DataStoreTransaction> runningTransactionMap = transactionRegistry.getRunningTransactions();
+            String filterByStatus = "status=in=(" + QueryStatus.CANCELLED.toString() + ","
+                    + QueryStatus.PROCESSING.toString() + ","
+                    + QueryStatus.QUEUED.toString() + ")";
+            FilterExpression fltStatusExpression = filterParser.parseFilterExpression(filterByStatus,
+                    AsyncQuery.class, false);
+            Collection<AsyncQuery> asyncQueryCollection =
+                    asyncQueryDao.getActiveAsyncQueryCollection(fltStatusExpression);
+
+            Set<UUID> runningTransactions = runningTransactionMap.keySet();
+
+            Set<UUID> asyncTransactions = asyncQueryCollection.stream()
+                    .filter(query -> query.getStatus() == QueryStatus.CANCELLED
+                    || TimeUnit.SECONDS.convert(Math.abs(query.getUpdatedOn().getTime()
+                    - new Date(System.currentTimeMillis()).getTime()), TimeUnit.MILLISECONDS) > maxRunTimeSeconds)
+                    .map(AsyncQuery::getRequestId)
+            .collect(Collectors.toSet());
+
+            Set<UUID> queriesToCancel = Sets.intersection(runningTransactions, asyncTransactions);
+
+            List<String> asyncQueryList = new ArrayList<String>();
+            queriesToCancel.stream()
+               .forEach((tx) -> {
+                   JsonApiDocument jsonApiDoc = new JsonApiDocument();
+                   MultivaluedMap<String, String> queryParams = new MultivaluedHashMap<String, String>();
+                   RequestScope scope = new RequestScope("query", NO_VERSION, jsonApiDoc,
+                           transactionRegistry.getRunningTransaction(tx), null, queryParams,
+                           tx, elide.getElideSettings());
+                   transactionRegistry.getRunningTransaction(tx).cancel(scope);
+                   List<String> cancelledQuery = asyncQueryCollection.stream().filter(query
+                           -> query.getRequestId().equals(tx)).map(AsyncQuery::getId).collect(Collectors.toList());
+                   asyncQueryList.addAll(cancelledQuery);
+               });
+            if (!queriesToCancel.isEmpty()) {
+                String cancelledQueryStr = String.join(",", asyncQueryList);
+                String filterById = "id=in=(" + cancelledQueryStr + ")";
+                FilterExpression fltIdExpression = filterParser.parseFilterExpression
+                        (filterById, AsyncQuery.class, false);
+                asyncQueryDao.updateStatusAsyncQueryCollection(fltIdExpression, QueryStatus.CANCEL_COMPLETE);
+            }
+        } catch (ParseException e) {
```

Comment:
The way the code is now, cancellation happens before we mark the status of the query as CANCEL_COMPLETE. So if there were to be an exception, the query status will not change, so it can be picked up for cancellation in the next run of the cancel thread...

Reply:
I was testing the code in example application and discovered that the executor does not schedule another thread if it encounters an exception. Can I catch exceptions, log them?